include/openthread/platform/crypto.h

```diff
@@ -0,0 +1,489 @@
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file includes the platform abstraction for Crypto operations.
+ */
+
+#ifndef OPENTHREAD_PLATFORM_CRYPTO_H_
+#define OPENTHREAD_PLATFORM_CRYPTO_H_
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <openthread/error.h>
+
+#ifdef OT_PLAT_CRYPTO_ATTRIBUTES_TYPE_HEADER
+#include OT_PLAT_CRYPTO_ATTRIBUTES_TYPE_HEADER
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup plat-crypto
+ *
+ * @brief
+ *   This module includes the platform abstraction for Crypto.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This enumeration defines the key types.
+ *
+ */
+typedef enum
+{
+    OT_CRYPTO_KEY_TYPE_RAW,  ///< Key Type: Raw Data.
+    OT_CRYPTO_KEY_TYPE_AES,  ///< Key Type: AES.
+    OT_CRYPTO_KEY_TYPE_HMAC, ///< Key Type: HMAC.
+} otCryptoKeyType;
+
+/**
+ * This enumeration defines the key algorithms.
+ *
+ */
+typedef enum
+{
+    OT_CRYPTO_KEY_ALG_VENDOR,       ///< Key Algorithm: Vendor Defined.
+    OT_CRYPTO_KEY_ALG_AES_ECB,      ///< Key Algorithm: AES ECB.
+    OT_CRYPTO_KEY_ALG_HMAC_SHA_256, ///< Key Algorithm: HMAC SHA-256.
+} otCryptoKeyAlgorithm;
+
+/**
+ * This enumeration defines the key usage flags.
+ *
+ */
+enum
+{
+    OT_CRYPTO_KEY_USAGE_NONE      = 0,      ///< Key Usage: Key Usage is empty.
+    OT_CRYPTO_KEY_USAGE_EXPORT    = 1 << 0, ///< Key Usage: Key can be exported.
+    OT_CRYPTO_KEY_USAGE_ENCRYPT   = 1 << 1, ///< Key Usage: Encryption (vendor defined).
+    OT_CRYPTO_KEY_USAGE_DECRYPT   = 1 << 2, ///< Key Usage: AES ECB.
+    OT_CRYPTO_KEY_USAGE_SIGN_HASH = 1 << 3, ///< Key Usage: HMAC SHA-256.
+};
+
+/**
+ * This enumeration defines the key storage types.
+ *
+ */
+typedef enum
+{
+    OT_CRYPTO_KEY_STORAGE_VOLATILE,   ///< Key Persistence: Key is volatile.
+    OT_CRYPTO_KEY_STORAGE_PERSISTENT, ///< Key Persistence: Key is persistent.
+} otCryptoKeyStorage;
+
+/**
+ * This datatype represents the key reference.
+ *
+ */
+typedef uint32_t otCryptoKeyRef;
```

Comment:
does this have to be uint32_t?

Reply:
@bukepo Yes, in PSA Crypto Key references are 32 bit unsigned ints.